### DISCUSSION TITLE
how to install a basic gcc toolchain?

### DISCUSSION THREAD

**Discussion by VladimirAlexiev** [ScoopInstaller/Scoop#5898] [Help and Questions] [answered]
(@rashil2000 please help)

I need a basic gcc toolchain (gcc, make, config) to be able to build Perl XS modules.
I installed `msys2, ming, gcc-msys, make, coreutils, binutils`: no luck.
Please see https://github.com/ScoopInstaller/Scoop/discussions/5889 for more details about my tribulations :-(

(I made a new question because the old one asked a different thing)

**Comment by VladimirAlexiev** [comment]
@rivy maybe you can help?
I see https://github.com/rivy/scoop-bucket/blob/master/perl.json ...

**Reply by rashil2000** [reply]
Uninstall any build-related tools that have been installed via Scoop. Install make, gcc, binutils, etc. all inside the MSYS2 MINGW64 shell. 

Unix-y tools installed independently via Scoop do not work very well.

**Comment by VladimirAlexiev** [comment, marked as answer]
I tried for an hour but faced various problems so I gave up and uninstalled msys2, mingw etc.

None of this is needed for Strawberry Perl, which has a minimal toolchain bundled.
Although `scoop install perl` also uses Strawberry, it omits this toolchain for some reason.
So I gave up and installed Strawberry from their site.

**Comment by rivy** [comment]
@VladimirAlexiev , what problems did you have with the referenced scoop recipe?

> Uninstall any build-related tools that have been installed via Scoop. Install make, gcc, binutils, etc. all inside the MSYS2 MINGW64 shell.
> 
> Unix-y tools installed independently via Scoop do not work very well.

Not really true.

I've got full toolchains for clang, gcc, perl, later versions of MSVC, python, and `make` all installed and working next to each other from my scoop variant and recipes. My current PATH currently even includes go, ruby, elixir, bcc32, embcc, and djgpp.

Notably, my PATH is setup with care: a basic system-wide path (`C:\WINDOWS\System32;C:\WINDOWS;C:\WINDOWS\System32\WindowsPowerShell\v1.0`), my local additions (`%USERPROFILE%\.local\bin;%USERPROFILE%\.deno\bin;%USERPROFILE%\.cargo\bin`), and then the scoop shim path (`%LOCALAPPDATA%\scoop\shims`). My scoop variant then takes explicit care to always add additional paths to PATH *after* the scoop shim path.

I just install the ezwinports `make` v4+ with additional aliases `gmake`, `gnumake`, and `gnu-make`. And I make sure that the MAKE environment variable is *undefined* (`perl` and `ruby`, specifically, and lots of Makefile recipes are dysfunctional if MAKE is set). 

I only rarely use the MSYS/MINGW shell, usually using WSL instead, if needed.

I'm not saying that everything will always work seamlessly, but when setup with a few recipe tweaks, I'm able to compile most things without any issues.

**Comment by VladimirAlexiev** [comment]
Hi @rivy! I haven't tried your recipe, should I do this to try it?
```
scoop bucket add rivy/scoop-bucket
```
My problem is that the standard scoop recipe installs Strawberry perl, but omits the toolchain bundled with it:
`c:\Strawberry\c\bin\*` includes gmake and bunch more (total 69 executables).
Does your recipe include a toolchain, or is there a recipe to install a toolchain with scoop?

```
c:\Strawberry\c\bin
# ls
addr2line.exe            libfribidi-0__.dll          libstdc++-6.dll                    xmlcatalog.exe
ar.exe                   libgcc_s_seh-1.dll          libtasn1-6.dll                     xmllint.exe
as.exe                   libgcrypt-20.dll            libtermcap-0__.dll                 xslt-config
c++.exe                  libgd-3__.dll               libtiff-6__.dll                    xsltproc.exe
c++filt.exe              libgdbm-6__.dll             libtiffxx-6__.dll                  xxhash.dll
ccache.exe               libgdbm_compat-4__.dll      libunistring-5.dll                 yasm.exe
cmcldeps.exe             libgfortran-5.dll           libunistring-5__.dll               ytasm.exe
cpack.exe                libgif-7__.dll              libwebp-7__.dll                    zlib1.dll
cpp.exe                  libgmp-10.dll               libwebpdemux-2__.dll               zlib1__.dll
ctest.exe                libgnutls-30.dll            libwebpmux-3__.dll
dlltool.exe              libgomp-1.dll               libwinpthread-1.dll
dllwrap.exe              libgomp-plugin-nvptx-1.dll  libxml2-2__.dll
dos2unix.exe             libgpg-error6-0.dll         libxslt-1__.dll
doxygen.exe              libgraphite2__.dll          libzstd.dll
elfedit.exe              libgss-3.dll                lto-dump.exe
freetype-config.bat      libharfbuzz-0__.dll         lz4-1.dll
g++.exe                  libharfbuzz-subset-0__.dll  mac2unix.exe
gcc-ar.exe               libhiredis.dll              mingw32-make.exe
gcc-nm.exe               libhistory8__.dll           mman-win.dll
gcc-ranlib.exe           libhogweed-6.dll            mysql_config.bat
gcc.exe                  libiconv-2.dll              nasm.exe
gcov-dump.exe            libiconv-2__.dll            ndisasm.exe
gcov-tool.exe            libidn-12.dll               ninja.exe
gcov.exe                 libidn2-0__.dll             nm.exe
gdb.exe                  libintl-8.dll               objcopy.exe
gendef.exe               libisl-23.dll               objdump.exe
genidl.exe               libjpeg-9__.dll             openssl.exe
gfortran.exe             liblzma-5.dll               patch.exe
gmake.exe                liblzma-5__.dll             pexports.exe
gprof.exe                liblzo2-2.dll               pg_config.exe
iconv.exe                libmpc-3.dll                ranlib.exe
jwasm.exe                libmpfr-6.dll               readelf.exe
ld.exe                   libmysql__.dll              size.exe
libXpm__.dll             libncursesw6.dll            strings.exe
libarchive-13.dll        libnettle-8.dll             strip.exe
libatomic-1.dll          libpng-config               unix2dos.exe
libbz2-1__.dll           libpng16-16__.dll           unix2mac.exe
libbz2.dll               libpng16-config             vsyasm.exe
libcharset-1__.dll       libpq__.dll                 widl.exe
libcrypto-1_1-x64__.dll  libquadmath-0.dll           windmc.exe
libcurl-4.dll            libreadline8__.dll          windres.exe
libdb-6.2__.dll          librhash.dll                x86_64-w64-mingw32-c++.exe
libdl.dll                librtmp-1.dll               x86_64-w64-mingw32-g++.exe
libexpat-1.dll           libsharpyuv-0__.dll         x86_64-w64-mingw32-gcc-13.1.0.exe
libexpat-1__.dll         libshishi-0.dll             x86_64-w64-mingw32-gcc-ar.exe
libexslt-0__.dll         libsource-highlight-4.dll   x86_64-w64-mingw32-gcc-nm.exe
libffi-6__.dll           libssh2-1.dll               x86_64-w64-mingw32-gcc-ranlib.exe
libfontconfig-1__.dll    libssh2-1__.dll             x86_64-w64-mingw32-gcc.exe
libfreeglut__.dll        libssl-1_1-x64__.dll        x86_64-w64-mingw32-gfortran.exe
libfreetype-6__.dll      libssp-0.dll                xml2-config
```

**Reply by rivy** [reply]
The download (from Strawberry) includes the toolchain.

The recipe I wrote revises the perl settings (in *portable.perl*) to point to the relevant toolchain utilities, duplicates any needed DLLs from the toolchain to the main perl directory, add then adds perl (via PATH) and both dmake and gmake (via shims).

I've been using it with success for years.

**Reply by VladimirAlexiev** [reply]
Thanks @rivy! Can you show a full list of things to install? First
```
scoop bucket add rivy/scoop-bucket  https://github.com/rivy/scoop-bucket
```
and then?
